File: crates/stylex-css-parser/src/css_types/dimension.rs
/*!
CSS Dimension type parsing.

Handles dimensional values that can be lengths, times, frequencies, or resolutions.
*/

use crate::{token_parser::TokenParser, token_types::SimpleToken};
use std::fmt::{self, Display};

use super::{frequency::Frequency, length::Length, resolution::Resolution, time::Time};

/// Union type for all dimensional CSS values
#[derive(Debug, Clone, PartialEq)]
pub enum Dimension {
  Length(Length),
  Time(Time),
  Frequency(Frequency),
  Resolution(Resolution),
}

impl Dimension {
  /// Check if a unit belongs to any dimension type
  #[allow(dead_code)]
  fn is_valid_dimension_unit(unit: &str) -> bool {
    Length::is_valid_unit(unit)
      || Time::is_valid_unit(unit)
      || Frequency::is_valid_unit(unit)
      || Resolution::is_valid_unit(unit)
  }

  /// Create a Dimension from value and unit
  fn from_value_and_unit(value: f32, unit: impl Into<String>) -> Option<Dimension> {
    let unit = unit.into();

    if Length::is_valid_unit(&unit) {
      Some(Dimension::Length(Length::new(value, unit)))
    } else if Time::is_valid_unit(&unit) {
      Some(Dimension::Time(Time::new(value, unit)))
    } else if Frequency::is_valid_unit(&unit) {
      Some(Dimension::Frequency(Frequency::new(value, unit)))
    } else if Resolution::is_valid_unit(&unit) {
      Some(Dimension::Resolution(Resolution::new(value, unit)))
    } else {
      None
    }
  }

  /// Parser for dimensional values
  pub fn parse() -> TokenParser<Dimension> {
    use crate::token_parser::tokens;

    tokens::dimension()
      .map(
        |token| {
          if let SimpleToken::Dimension { value, unit } = token {
            Self::from_value_and_unit(value as f32, unit)
          } else {
            None
          }
        },
        Some("extract_dimension"),
      )
      .where_fn(|opt| opt.is_some(), Some("valid_dimension"))
      .map(|opt| opt.unwrap(), Some("unwrap_dimension"))
  }
}

pub fn dimension() -> TokenParser<Dimension> {
  Dimension::parse()
}

#[cfg_attr(coverage_nightly, coverage(off))]
impl Display for Dimension {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    match self {
      Dimension::Length(length) => length.fmt(f),
      Dimension::Time(time) => time.fmt(f),
      Dimension::Frequency(frequency) => frequency.fmt(f),
      Dimension::Resolution(resolution) => resolution.fmt(f),
    }
  }
}

#[cfg(test)]
#[path = "../tests/css_types/dimension_tests.rs"]
mod tests;

#[cfg(test)]
#[path = "../tests/css_types/dimension_test.rs"]
mod dimension_test;

File: packages/style-value-parser/lib/css-types/dimension.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.dimension = void 0;
var _tokenParser = require("../token-parser");
var _frequency = require("./frequency");
var _length = require("./length");
var _resolution = require("./resolution");
var _time = require("./time");
function arrIncludes(arr, val) {
  return arr.includes(val);
}
const dimension = exports.dimension = _tokenParser.TokenParser.tokens.Dimension.map(token => {
  const {
    unit,
    value
  } = token[4];
  if (arrIncludes(_length.Length.UNITS, unit)) {
    return new _length.Length(value, unit);
  } else if (arrIncludes(_time.Time.UNITS, unit)) {
    return new _time.Time(value, unit);
  } else if (arrIncludes(_resolution.Resolution.UNITS, unit)) {
    return new _resolution.Resolution(value, unit);
  } else if (arrIncludes(_frequency.Frequency.UNITS, unit)) {
    return new _frequency.Frequency(value, unit);
  } else {
    null;
  }
}).where(val => val != null);